File: src/urich/core/routing.py
"""Route module: one object per context; routes are attached to it."""
from __future__ import annotations

from typing import Any, Callable

from urich.core.app import Application
from urich.core.module import Module


class HttpModule(Module):
    """
    HTTP module (bounded context): name + routes.
    Attach via app.register(module). Similar to include_router in FastAPI.
    """

    def __init__(self, name: str, prefix: str | None = None) -> None:
        self.name = name
        self.prefix = prefix or f"/{name}"
        self._routes: list[tuple[str, Any, list[str]]] = []

    def route(self, path: str, endpoint: Callable[..., Any], methods: list[str] | None = None) -> HttpModule:
        """Add a route. path without leading slash is under the module prefix."""
        if methods is None:
            methods = ["GET"]
        p = path if path.startswith("/") else f"/{path}"
        self._routes.append((p, endpoint, methods))
        return self

    def register_into(self, app: Application) -> None:
        for path, endpoint, methods in self._routes:
            full_path = self.prefix.rstrip("/") + path
            app.add_route(full_path, endpoint, methods)

File: urich-rs/src/core/routing.rs
//! HttpModule: route module by prefix. Like Python core/routing.

use std::sync::Arc;
use urich_core::CoreError;

use super::app::Application;
use super::{Handler, Module};

/// HTTP module (bounded context): name + routes. Like Python HttpModule.
/// Attach via app.register(module). Similar to include_router in FastAPI.
pub struct HttpModule {
    pub name: String,
    pub prefix: String,
    routes: Vec<(String, Arc<dyn Fn(serde_json::Value) -> Result<serde_json::Value, CoreError> + Send + Sync>, String)>,
}

impl HttpModule {
    pub fn new(name: &str, prefix: Option<&str>) -> Self {
        Self {
            name: name.to_string(),
            prefix: prefix.unwrap_or(&format!("/{}", name)).to_string(),
            routes: Vec::new(),
        }
    }

    /// Add a route. path without leading slash is under the module prefix.
    /// methods e.g. ["GET"], ["GET", "POST"].
    pub fn route(
        mut self,
        path: &str,
        handler: impl Fn(serde_json::Value) -> Result<serde_json::Value, CoreError> + Send + Sync + 'static,
        methods: &[&str],
    ) -> Self {
        let full_path = format!(
            "{}/{}",
            self.prefix.trim_end_matches('/'),
            path.trim_start_matches('/')
        );
        let arc = Arc::new(handler);
        for method in methods {
            self.routes
                .push((full_path.clone(), arc.clone(), method.to_string()));
        }
        self
    }
}

impl Module for HttpModule {
    fn register_into(&mut self, app: &mut Application) -> Result<(), CoreError> {
        for (path, arc, method) in self.routes.drain(..) {
            let handler: Handler = Box::new(move |v| arc(v));
            app.register_route(&method, &path, None, handler, None)?;
        }
        Ok(())
    }
}
